Update GPIO API to support flags used by Linux DT bindings

mnkp opened the pull request:
This is a first PR in series which updates GPIO API, drivers and samples to implement changes discussed in #15611, #10339. It is based on work done in #12651, #11880.

The primary focus of this PR is to introduce standard flags used by [Linux DT bindings](https://elixir.bootlin.com/linux/latest/source/include/dt-bindings/gpio/gpio.h). The specific changes implemented:
- Added GPIO_ACTIVE_LOW, GPIO_ACTIVE_HIGH to indicate pin active state.
- Added GPIO_OPEN_DRAIN, GPIO_OPEN_SOURCE to configure single ended pin
  driving mode.
- Added GPIO_PULL_UP, GPIO_PULL_DOWN flags.

Support for GPIO_PERSISTENT, GPIO_TRANSITORY flags (bit 3) will be added later.

Remaining changes:
- Added flags to support [gpio](https://www.kernel.org/doc/Documentation/devicetree/bindings/gpio/gpio.txt) and [pinctrl](https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pinctrl-bindings.txt) bindings defined in form of properties.
    - GPIO_INPUT, GPIO_OUTPUT to configure pin as input or output.
    - Added GPIO_OUTPUT_LOW, GPIO_OUTPUT_HIGH flags to initialize output in low or high state.

- All GPIO flags which should not be used in DTS were removed from DT bindings gpio.h file. Support for DTS properties of GPIO node such as: input, output-low, output-high, line-name, etc. will be added in the future.
- GPIO_INT_* flags were reworked to support configuration of pin interrupts on logical or physical levels.
- Following flags were deprecated: GPIO_DIR_*, GPIO_DS_DISCONNECT_*, GPIO_PUD_*, GPIO_INT_ACTIVE_*, GPIO_INT_DOUBLE_EDGE, GPIO_POL_*.
- Added gpio_pin_set, gpio_pin_get functions which work with pin logical levels, taking into account GPIO_ACTIVE_LOW flag

gpio_sam, gpio_gecko drivers along with .dts and other infrastructure files were updated to reflect above changes. Also 'button' and 'blinky' samples were updated to use the new GPIO flags.

This PR is not yet complete. To be able to merge it we need to update all existing GPIO drivers, samples and tests to use the new flags.

Any functional changes not mentioned here will be addressed in subsequent PRs. Specifically this PR doesn't modify flags related to pin drive strength, debouncing neither attempts to introduce functions operating on ports.

mnkp committed:
api: gpio: Align GPIO dt-bindings flags with Linux DTS

This commit makes following changes to GPIO dt-bindings flags:
- Added GPIO_ACTIVE_LOW, GPIO_ACTIVE_HIGH to indicate pin active state.
- Added GPIO_OPEN_DRAIN, GPIO_OPEN_SOURCE to configure single ended pin
  driving mode.
- Added GPIO_PULL_UP, GPIO_PULL_DOWN flags.
- GPIO_INPUT, GPIO_OUTPUT to configure pin as input or output.
- Added GPIO_OUTPUT_LOW, GPIO_OUTPUT_HIGH flags to initialize output
  in low or high state.
- reworked GPIO_INT_* flags to configure pin interrupts.
- following flags were deprecated: GPIO_DIR_*, GPIO_DS_DISCONNECT_*,
  GPIO_PUD_*, GPIO_INT_ACTIVE_*, GPIO_INT_DOUBLE_EDGE, GPIO_POL_*.

To be aligned with Linux DTS standard any GPIO flags that should not be
used in DTS files are moved from include/dt-bindings/gpio/gpio.h file to
include/drivers/gpio.h with an exception of several old flags which
removal would cause DTS compilation errors. Those remaining old flags
will be removed from include/dt-bindings/gpio/gpio.h at a later stage.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
drivers: gpio_gecko: update to use new GPIO API

Update driver code and board files to use new GPIO configuration flags
such as GPIO_ACTIVE_LOW. Also add implementation of new port_* driver
API as well as gpio_pin_interrupt_configure function.

Tested on efr32_slwstk6061a board.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
samples: blinky: update to use new GPIO API

Update gpio_pin_configure() to take into account GPIO flags defined by
the devicetree.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
drivers: gpio_sam: update to use new GPIO API

Update driver code and board files to use new GPIO configuration flags
such as GPIO_ACTIVE_LOW. Also add implementation of new port_* driver
API as well as gpio_pin_interrupt_configure function.

Tested on sam_e70_xplained board.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
gpio: add new functions to set/get pin/port values

This commit adds following functions which work with pin logical levels
(take into account GPIO_ACTIVE_LOW flag):
- gpio_port_get, gpio_port_set_masked, gpio_port_set_bits,
  gpio_port_clear_bits, gpio_port_set_clr_bits
- gpio_pin_get, gpio_pin_set
Functions which work with pin physical levels:
- gpio_port_get_raw, gpio_port_set_masked_raw, gpio_port_set_bits_raw,
  gpio_port_clear_bits_raw, gpio_port_set_clr_bits_raw
- gpio_pin_get_raw, gpio_pin_set_raw
As well as functions:
- gpio_port_toggle_bits, gpio_pin_toggle_bits

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
samples: button: update to use new GPIO API

Update gpio_pin_configure() to take into account GPIO flags defined by
the devicetree. Use gpio_pin_get/gpio_pin_set to verify reading/writing
of logical pin values. Use gpio_pin_interrupt_configure() to configure
interrupts.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mbolivar commented on drivers/gpio/gpio_handlers.c:
```diff
@@ -27,6 +27,18 @@ Z_SYSCALL_HANDLER(gpio_read, port, access_op, pin, value)
+Z_SYSCALL_HANDLER(gpio_pin_set, port, pin, value)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_GPIO(port, pin_set));
+	return z_impl_gpio_write((struct device *)port, pin, value);
```
Shouldn't this be calling z_impl_gpio_pin_set?

mbolivar commented on drivers/gpio/gpio_handlers.c:
```diff
@@ -27,6 +27,18 @@ Z_SYSCALL_HANDLER(gpio_read, port, access_op, pin, value)
+Z_SYSCALL_HANDLER(gpio_pin_get, port, pin)
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_GPIO(port, pin_get));
+	return z_impl_gpio_write((struct device *)port, pin);
```
And this one z_impl_gpio_pin_get?

mbolivar reviewed (commented):
This review is for the API changes only. I didn't have time time to look at the samples or driver changes.

mnkp commented on include/dt-bindings/gpio/gpio.h:
```diff
@@ -6,102 +6,177 @@
-/** GPIO pin trigger on level low or falling edge. */
-#define GPIO_INT_ACTIVE_LOW	(0 << 2)
+/** Enables GPIO pin pull-up. */
+#define GPIO_PULL_UP            (1 << 4)
```
Bit 3 is defined by [Linux DT bindings](https://elixir.bootlin.com/linux/latest/source/include/dt-bindings/gpio/gpio.h) as:
```
/* Bit 3 express GPIO suspend/resume and reset persistence */
#define GPIO_PERSISTENT 0
#define GPIO_TRANSITORY 8
```
We should add support for this flag later. I updated issue description to clarify this. Bits 6, 7 are reserved in case Linux guys add more flags.

Thanks for the feedback! I fixed all the other issues you reported.

mbolivar replied:
Ah, I get it now; I didn't realize you were matching values a well as symbolic names. Thanks for explaining.

mbolivar replied:
> Thanks for the feedback! I fixed all the other issues you reported.

Just FYI, it took me a while to make sure that this was true, because the fix was squashed into f5857b7645 ("api: move non DTS GPIO flags away from DT bindings"), which was just supposed to be a copy/paste without changes type commit; I would have expected it in ("f64993a3ea api: gpio: Clean up GPIO dt-bindings flags")

mbolivar commented on drivers/gpio/gpio_gecko.c:
```diff
@@ -91,18 +93,31 @@ static int gpio_gecko_configure(struct device *dev,
 
-	if ((flags & GPIO_DIR_MASK) == GPIO_DIR_IN) {
-		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
+	if (flags & GPIO_OUTPUT) {
```
When I was looking at the flags, I wondered what the code would do when both GPIO_OUTPUT and GPIO_INPUT were set.

This driver respects output before input, but I don't see it documented that OUTPUT always has higher priority if both are set. Is that your intent?

I think it should be documented if so. Though my preference would be slightly for rejecting this as an invalid combination -- perhaps a common helper function, `gpio_flags_are_valid(u32_t flags)`, could be added for this purpose?

pabigot replied:
There are several devices where it's perfectly valid and very useful to have a particular GPIO configured for input and output simultaneously.  The API should allow that, on those devices.

mbolivar replied:
Fair enough. What are the intended semantics if the user requests both but the driver only supports one?

pabigot replied:
Dunno; `-ENOTSUP` would seem to make sense, though `-EINVAL` is used here for a combination that would be acceptable on other platforms (`GPIO_INT` with `GPIO_OUTPUT`should be fine if `GPIO_INPUT` is also set).

In this case the comments suggest that bidirectional is handled in `GPIO_OUTPUT` case, though I don't see how that works and it's not how I would expect it to be expressed for Nordic or SX1509B.  In those input and output are simply independent functions.

The API documentation needs to be reviewed and updated to describe exactly what is and is not allowed, and what behavior the different combinations have to provide in order to meet the API's promises.  `GPIO_DEBOUNCE` is one I noticed that still has about the same chance for misinterpretation as `GPIO_POL_INV` did in the current API.

mbolivar replied:
`-ENOTSUP` sounds like a good plan to me for unsupported (but valid) combinations of flags.

mnkp replied:
This bit of code was written specifically for the gecko GPIO module. In case any of the output modes are enabled the input will also be implicitly enabled. This is actually a fairly common approach among SoC vendors but surely not the only one. Application code that wants to read from the input must pass `GPIO_INPUT` flag to the configure function. In case of the gecko platform if it passes `GPIO_OUTPUT` only the input direction will still work. A portable application code shouldn't rely on this 'feature'. The gecko GPIO driver doesn't perform extra checks, e.g. the implementation of `gpio_pin_get` function doesn't check if the input direction was explicitly configured as this would increase function's code size, make it slower.

As for the `GPIO_OUTPUT`, `GPIO_INPUT` flag usage. If an application wants to write to the pin it should pass `GPIO_OUTPUT` flag to the configure function, if it intends to read from the pin it should pass `GPIO_INPUT` flag. If it intends to write and read it should pass both flags. If no flag is passed the pin will be disabled.

In general, if the function is not able to provide features requested by the configuration flags it should return `-ENOTSUP`.

I agree though we should better document error code / assert usage.

mbolivar replied:
> I agree though we should better document error code / assert usage.

That would be great if you don't mind. I really like your proposal of adding asserts into gpio.h itself in addition to docs. :+1:

mbolivar commented on drivers/gpio/gpio_gecko.c:
```diff
@@ -91,18 +93,31 @@ static int gpio_gecko_configure(struct device *dev,
 			mode = gpioModeInputPull;
 			out = 1U; /* pull-up*/
-		} else if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_DOWN) {
+		} else if (flags & GPIO_PULL_DOWN) {
```
Similar comment here. I think pull up and pull down both set should result in -EINVAL. Is there a good reason not to?

In general, if one flag is meant to take precedence over another, I think some static inline helpers like `gpio_flags_pull_up()`, `gpio_flags_pull_down()`, etc. should be added which respect the precedence ordering, which can be called from all drivers. This will ensure consistent behavior when moving programs between SoCs. (But I still personally think -EINVAL is a little better.)

mbolivar commented on drivers/gpio/gpio_sam.c:
```diff
@@ -22,64 +22,109 @@ struct gpio_sam_config {
 
 	/* Setup Pull-up resistor. */
-	if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
+	if (flags & GPIO_PULL_UP) {
```
Does it really make sense to let the user turn on both the pull up and pull down resistors? Sorry if I am not understanding the idea, but it doesn't seem sensible.

I guess part of the Zephyr philosophy is "minimal error checking". But maybe we could add some asserts for inconsistent behavior like this if you agree it doesn't make sense, but don't want to add code or runtime cost to check this sort of thing in production builds.

mnkp replied:
I don't think there is any real life use case where an application may want to enable both pull up and pull down resistors even if some hardware will happily do it. I don't perform checks not to add extra run time code but indeed we could handle this with an assert. I'll add it in include/gpio.h file.

mbolivar reviewed (commented):
Thank you for the updates! I've had some time to look at the use in drivers and applications and have some questions.

Keep up the great work!

mbolivar reviewed (dismissed):
requesting changes while semantic specification debate is ongoing

b0661 reviewed (commented):
I think to have a single function to configure GPIO interrupts and GPIO I/O characteristics is a misconception. There is a reason why Linux only provides GPIO DT bindings for the latter.

In my use case I have a SPI chip select handled by hardware (pin muxed to SPI) where I still want to get interrupts on an active chip select. If I would configure the interrupt by gpio_pin_configure() the pin would be muxed to the GPIO function.

Interrupt configuration should be done in the gpio_callback struct. (E.g. in STM32 interrupt generation and GPIO are even two distinct blocks.)

pabigot commented:
> Interrupt configuration should be done in the gpio_callback struct. (E.g. in STM32 interrupt generation and GPIO are even two distinct blocks.)

Without deep analysis this makes a lot of sense.  Some drivers would have to make sure they preserve non-interrupt-related configuration as interrupts are enabled and disabled, but that shouldn't be too painful.

Depending on the model used to describe interrupt vs gpio configurations I can imagine that for some hardware the GPIO might have to be changed during the periods when interrupts are enabled (e.g. enable `GPIO_INPUT` to receive the signals), but should be restored when they're disabled (e.g. disable `GPIO_INPUT` for power savings).   This could introduce a need for the driver to maintain pin-specific state internally rather than rely on what can be read back from peripheral registers.  I don't have a problem with that in principle, and in fact it would make it trivial to add API to read the current GPIO configuration, something I've sometimes wanted to do.

anangl reviewed (commented):
LGTM.

My little doubt is only regarding flags with value of 0, like `GPIO_INT_LEVELS_PHYSICAL` or `GPIO_INT_LEVEL`.
Without checking the actual value of such flag, perhaps when modifying some code, one could try to use something like `if (flags & GPIO_INT_LEVELS_PHYSICAL)`, which obviously won't work as expected. We need to support the flags of value 0 defined in DT bindings, but maybe we could get rid of these two (not add `GPIO_INT_LEVELS_PHYSICAL` and deprecate `GPIO_INT_LEVEL`)? Are they actually useful?

And I support the @b0661's proposal of moving the interrupt configuration flags to the callback structure. This would also allow creating generic parts of code that could enable/disable/reconfigure interrupts without having to know the actual pin configuration.

mnkp committed:
gpio: add gpio_pin_interrupt_configure function

This commit moves interrupt configuration for a single pin from
gpio_pin_configure to gpio_pin_interrupt_configure function.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp commented:
> My little doubt is only regarding flags with value of 0, like `GPIO_INT_LEVELS_PHYSICAL` or `GPIO_INT_LEVEL`.
> Without checking the actual value of such flag, perhaps when modifying some code, one could try to use something like `if (flags & GPIO_INT_LEVELS_PHYSICAL)`, which obviously won't work as expected.

That's indeed a limitation of this approach. When writing a device driver one needs to always check the definition of the flag not to use a 'dummy' one. They shouldn't be used by the device drivers, only by the application level code. And in this case they make the code easier to read. The application level code will never want to perform any operation on these flags so there is no danger that the flag will be misused. It can be misused by the author of the device driver but this is entirely under our control, happens only once when writing the driver and should be caught when testing the code. I removed `GPIO_INT_LEVELS_PHYSICAL` flag but would prefer to leave the other flags as is.

mnkp commented:
I've added `gpio_pin_interrupt_configure` which can be used to independently configure pin interrupts. I'll provide more rationale for this approach in #17102 issue.

The `gpio_pin_interrupt_configure` flags parameter is an `unsigned int` while in case of `gpio_pin_configure` the flags parameter is `int`. I'll fix the latter but for the moment didn't want to create 'noise' since all the gpio drivers need to be updated.

I've started to improve description of the errors various functions can return. This is not yet finished.

pabigot commented on include/drivers/gpio.h:
```diff
@@ -32,6 +32,320 @@ extern "C" {
+ *
+ * The interface supports two different drive strengths:
+ * `DFLT` - The lowest drive strength supported by the HW
+ * `ALT` - The highest drive strength supported by the HW
```
Nordic and SX1509B differ in whether the peripheral-defined default is lower or higher strength than the alternative setting.   On Nordic it's standard and high; on SX1509B it's normal and reduced.  I suppose it's prior art (introduced for Nordic) so you have to do something, but IMO this is the sort of thing that doesn't belong in a generic driver API.  There's really no way this can be used portably.

mnkp replied:
As I mentioned in the PR description
> Any functional changes not mentioned here will be addressed in subsequent PRs. Specifically this PR doesn't modify flags related to pin drive strength, debouncing neither attempts to introduce functions operating on ports.

Linux DTS nodes define drive strength by specifying sink or source current in mA. To have compliant DTS description we would need to do the same, but maybe it's not worth it. I proposed to handle this issue later. It has no influence on the modifications introduced by this PR.

pabigot commented:
> And in this case they make the code easier to read.

IMO it's more readable only if it's correct and the reviewer is trusting.  Every time I see `flags & FOO` I feel obliged to go look at the header to see if it's correct, and that it's still (and truly) a single-bit flag (i.e. `FOO` isn't a helper macro that aggregates a bunch of bits, and the condition succeeds because of a match in a subordinate bit).  With `(flags & FOO_Mask) == FOO_value` I don't worry so much.

>  The application level code will never want to perform any operation on these flags so there is no danger that the flag will be misused.

I'm not convinced this assertion is justified.  I can certainly imagine situations where I'm providing a helper that can see the flags and needs to inspect them to perform its task.

> It can be misused by the author of the device driver but this is entirely under our control, happens only once when writing the driver and should be caught when testing the code.

I have not found this to be the case.  At least two errors of this sort were identified and corrected in patches associated with my attempt at this task.

pabigot commented on include/drivers/gpio.h:
```diff
@@ -233,6 +543,75 @@ static inline int gpio_pin_read(struct device *port, u32_t pin,
+ *
+ * @param port Pointer to device structure for the driver instance.
+ * @param pin Pin number to configure.
+ * @param flags Interrupt configuration flags as defined by GPIO_INT_*.
```
My first concern here is that this suggests that only flags that begin with `GPIO_INT_*` can appear in the `flags` parameter.  This implies that those flags should *not* appear in the `flags` parameter to `gpio_pin_configure`, but that is not documented.

The problem is that when dealing with a pin that is used both in non-interrupt (output chip-select, input polling) and an interrupt (device needs service) I can imagine situations where other flags might be desired.  A specific example is that I might need a pull-up on an active-low interrupt, but it's not part of the gpio configuration flags for use as an output.

So: which flags are permitted in which API?

Also, in the telecon we discussed whether configuring a pin for interrupt use should block use of the gpio API on it (e.g. you can't write to a pin that's used for interrupts).  I don't see any documentation on how interoperability between the two modes, nor implementation that prevents some operations.  (At a minimum enabling callbacks on a pin that's not configured for interrupts should be rejected, or something.)

Finally, this configures (essentially enables) interrupts separately from the callbacks, but doesn't take into account discussion from #17102 which proposes combining the configure, enable, and callback configuration into a single function that places a (or several) pins into interrupt mode.  I don't think this function should be added without considering how it will interact with that sort of refactoring (which, incidentally, would resolve some of the questions above).

mnkp replied:
I've pushed a new gpio.h version where description of flags parameter explicitly mentions which flags are allowed. Also added a comment clarifying that `gpio_pin_configure` should always be called first. Any other function in the GPIO API, including `gpio_pin_interrupt_configure` should be called only after the main pin configuration has been done. `gpio_pin_interrupt_configure` doesn't introduce a new working mode, only adds interrupt capability on top of an already existing pin configuration. If pin is configured as an output only the `gpio_pin_interrupt_configure` should report it via assert. The description of what checks are performed via assert is not yet added. I will do it next.

b0661 replied:
> Any other function in the GPIO API, including gpio_pin_interrupt_configure should be called only after the main pin configuration has been done.

I assume gpio_pin_configure ("main pin configuration") muxes the pin to GPIO. Requiring this before you can use gpio_pin_interrupt_configure prevents to set interrupt callbacks on pins that are muxed to alternate functions as SPI, I2C, ... In my view a showstopper. This is also an indication that the concept of interrupt on pin level changes and pin characteristic configuration should be handled independently (see e.g. https://elinux.org/images/b/b6/Pin_Control_Subsystem_Overview.pdf page 17/18). 

> If pin is configured as an output only the gpio_pin_interrupt_configure should report it via assert.

This should not be a general rule. There are SoCs that allow to generate an interrupt even when configured as an output.

mnkp replied:
The general idea is that the future pinctrl API will allow application to connect peripheral to a pin as well as configure pin properties: drive strength, pull up/down, open source drive mode, etc. The application will not need GPIO module to do that. I was planning to use the same approach for the interrupt configuration but actually interrupts are different. It makes sense to enforce application to configure interrupts via GPIO module. In this case the statement
> Any other function in the GPIO API, including gpio_pin_interrupt_configure should be called only after the main pin configuration has been done.

is not valid. I'll modify it to clarify that `gpio_pin_interrupt_configure()` can be called on the pin controlled by a peripheral module other than GPIO. Only documentation will change. The implementation stays as it is now.

Modifying properties of a pin controlled by a peripheral module other than GPIO will still be possible only after we introduce pinctrl driver.

> > If pin is configured as an output only the gpio_pin_interrupt_configure should report it via assert.
> 
> This should not be a general rule. There are SoCs that allow to generate an interrupt even when configured as an output.

Could you give an example of such a SoC? Please notice that my comment states "output only". Having a pin configured as input-output (or input) is fine. Typically, if the input path is disconnected in GPIO hardware module there is no way to generate interrupts as interrupt subsystem is blind, unaware of pin changes.

b0661 replied:
> Could you give an example of such a SoC? Please notice that my comment states "output only". Having a pin configured as input-output (or input) is fine.

Sorry, I misinterpreted your statement "output **only**" and took it for "output" without the input disconnected.

pabigot reviewed (dismissed):
This is a micro-review focused solely on the proposed `gpio_pin_interrupt_configure` API, to be dismissed when its questions have been resolved and documentation updated.

Also, please rebase this PR.  The samples have non-trivial conflicts with subsequent changes in master so it's hard to see how the API change will affect real programs.

b0661 commented:
You are creating 6 functions to manage the callback on interrupt. 
```
	int (*pin_interrupt_configure)(struct device *port, u32_t pin,
				       unsigned int flags);
	int (*pin_interrupt_enable)(struct device *port, u32_t pin);
	int (*pin_interrupt_disable)(struct device *port, u32_t pin);
	int (*manage_callback)(struct device *port, struct gpio_callback *cb,
			       bool set);
	int (*enable_callback)(struct device *port, int access_op, u32_t pin);
	int (*disable_callback)(struct device *port, int access_op, u32_t pin);
```

I think the same functionality can be achieved with a more minimal API:
```
struct gpio_callback {
	/** Actual callback function being called when relevant. */
	gpio_callback_handler_t handler;
	/** A mask of pins the callback is interested in.
	 * Callback is called if one of the pins is triggered according to pin_trigger_mode.
	 */
	u32_t pin_mask;
	/** Trigger mode of the pins the callback is interested in.  */
	u8_t pin_trigger_mode;
};
int (*enable_callback)(struct device *port, const struct gpio_callback *cb);
int (*disable_callback)(struct device *port, const struct gpio_callback *cb);
```

* pin_interrupt_configure -> enable_callback
* pin_interrupt_enable -> enable_callback
* pin_interrupt_disable -> disable_callback
* manage_callback -> enable_callback
* enable_callback -> enable_callback
* disable_callback -> disable_callback

Complexity of the driver stays the same as all the functionality is anyway needed. Benefit is that you may add several callbacks to a pin change and that the driver always knows about all callbacks (and the trigger_modes) for the pins. Misconfiguration or impossible configuration due to another callback requesting an incompatible trigger_mode can easily be detected. cb is a pointer to const to enable the callback struct data in Flash/ROM.

pabigot commented:
> You are creating 6 functions to manage the callback on interrupt.

Second.  That was the whole point of creating #17102 and #15611: so we could discuss the desired API *before* it was implemented, so when we review we all know what the goal was, and the PR comments can be on the PR, not bikeshedding the API.

We could do the design work here.  But it's a lot harder to find things.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -89,35 +376,27 @@ struct gpio_callback {
+	int (*pin_get_raw)(struct device *port, u32_t pin);
+	int (*pin_interrupt_configure)(struct device *port, u32_t pin,
+				       unsigned int flags);
+	int (*pin_interrupt_enable)(struct device *port, u32_t pin);
```
I propose to introduce for now only the `pin_*` versions of different API functions exported by the GPIO drivers. Some of the functions could be replaced by the version operating on the mask parameter and not a pin but that would require a larger modifications to the underlying driver code. Several of the HALs used by the drivers provide only pin versions of GPIO functions.

I propose to clean it up / use functions operating on the mask parameter later when we introduce GPIO `port_*` functions. I find it more convenient to split the work this way.

mnkp commented:
I've provided rationale for proposed changes to the interrupt configuration in #17102. This PR is only to showcase the resulting implementation. Let's continue the discussion related to the GPIO interrupt configuration in #17102.

mnkp committed:
drivers: gpio_nrfx: update to use new GPIO API

Update driver code and board files to use new GPIO configuration flags
such as GPIO_ACTIVE_LOW. Also add implementation of new port_* driver
API as well as gpio_pin_interrupt_configure function.

Tested on nrf52840_pca10056 board.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>
Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

mnkp committed:
drivers: gpio_mcux: update to use new GPIO API

Update driver code and board files to use new GPIO configuration flags
such as GPIO_ACTIVE_LOW. Also add implementation of new port_* driver
API as well as gpio_pin_interrupt_configure function.

Tested on frdm_k64f board.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

anangl commented on boards/arm/nrf9160_pca10090/nrf9160_pca10090_common.dts:
```diff
@@ -18,39 +18,39 @@
 		led0: led_0 {
-			gpios = <&gpio0 2 GPIO_INT_ACTIVE_HIGH>;
-			label = "Green LED 0";
+			gpios = <&gpio0 2 0>;
```
Why not `GPIO_ACTIVE_HIGH`? I thought that was the purpose of these flags with value of 0 - to use them to clearly express what is being configured.

mnkp replied:
Yes, we could add `GPIO_ACTIVE_HIGH` flag here. But to be consistent we would need to add it also to all other DTS files. So far we've been using flags to indicate that settings are different than default, i.e. without flags the pin defaults to active high and output defaults to push pull.

mnkp commented on boards/arm/nrf52840_pca10090/nrf52840_pca10090.dts:
```diff
@@ -21,6 +21,58 @@
 		zephyr,flash = &flash0;
 	};
+
+	leds {
```
My mistake, the data are for nRF9160 DK PCA 10090 board. I'll fix it.

mbolivar commented:
Dismissed my stale review

mnkp commented:
Due to several people asking to provide _port versions of functions controlling pin values I've introduced them in the latest update to this PR. Driver API (as defined by `struct gpio_driver_api`) doesn't include pin version of those functions any more instead pin functions are implemented directly in `include/drivers/gpio.h` based on port versions provided by the driver API. I've reviewed most, though not all, GPIO drivers available currently in Zephyr and it seems like this approach could work. It is a bit risky though since we assume that for all current and future architectures the correspondence between pin index and its representation in the port is given by (1 << `pin`) where `pin` is the pin index and that HAL libraries used by the drivers will provide required functions to manipulate port values. If necessary I will reintroduce pin functions directly to the driver API (as defined by `struct gpio_driver_api`).

mnkp commented:
Before pushing this PR on topic-gpio branch we need to implement proper handling of `__DEPRECATED_MACRO` by DTS compiler.

pabigot commented:
I don't see in the current PR the changes described [here](https://github.com/zephyrproject-rtos/zephyr/pull/16648#discussion_r310359012) or [here](https://github.com/zephyrproject-rtos/zephyr/pull/16648#discussion_r310540575) clarifying the use of flags in pin and interrupt configuration, and the coordination between pin and interrupt configuration.  Did they get lost?

galak commented:
> Before pushing this PR on topic-gpio branch we need to implement proper handling of `__DEPRECATED_MACRO` by DTS compiler.

this is highly unlikely to happen due to the fact that DTC doesn't support any such functionality and any change to support something would mean a new DTC version, a new SDK, etc. etc.

I think the best I've come up with so far, is that we split out the old dt-binding/gpio/gpio.h into a gpio-old.h and have a #warning if that's included.  We change all .dts{i} files to use gpio-old.h for now and as the SoC/drivers get updated we switch them to use gpio.h.

pabigot reviewed (changes requested):
I'm mostly OK with this for now, with the understanding that we'll all be proposing updates as we gain experience.

The one change I do want made before merge is in the separate comment in this review: make the `set_clear` API support toggling consistent with what we'd discussed.

As a heads-up here are things I'd like done in the topic branch before we go too far along in updating drivers.  From my perspective they are not blockers for merging this PR, and I'm willing to do the PRs to address them.  I just want the thoughts out there so people can think about them.

First I'd like to see some cleanup of the three visible API functions that are not documented (`gpio_config`, `gpio_write`, and `gpio_read`). If users aren't supposed to invoke them, can that be indicated in some way? Maybe a `z_` prefix?

Also in the topic branch before everybody starts working on drivers: we need a clear description of *all* the configuration flags including the ones not touched in this PR, such as `GPIO_INT_DEBOUNCE`, so implementers can determine whether the required functionality can be implemented.  For SX1509B the drive strength flags also need to be refined because the existing description is simply wrong for them.  I'd like code-owners for drivers that do implement these extra features to contribute to specifying the effect they must have.

~From earlier discussion on future-proofing we should `typedef u32_t gpio_pins_t` and use that for all parameters where the argument encodes a set of pins.  The documentation for the type alias should specify that pin `i` is in the set if and only if bit `i` (masked by `1U << i`) is 1. This would allow for a 64-bit pin mask or some more complex encoding if necessary; but more importantly it would make it far more clear to a user how to construct the corresponding argument.~ partially addressed in revisions.

I would also like to investigate the opportunities to reduce the number of API functions the driver must implement by delegating more things to the fundamental `set_clear` operator.

mnkp commented:
> The one change I do want made before merge is in the separate comment in this review: make the `set_clear` API support toggling consistent with what we'd discussed.

As far as I remember the last statement regarding `set_clear` toggling behavior that you made on the Slack #gpio channel was to "please forget that I proposed the toggling" or something among this lines. Since we have a dedicated `gpio_port_toggle_bits()` function this indeed seemed unnecessary. Nevertheless, I can add toggling, that's not a problem.

> First I'd like to see some cleanup of the three visible API functions that are not documented (`gpio_config`, `gpio_write`, and `gpio_read`). If users aren't supposed to invoke them, can that be indicated in some way? Maybe a `z_` prefix?

`gpio_config`, `gpio_write`, and `gpio_read` are specifically marked as for internal use only. I could add deprecate macro for `gpio_write`, and `gpio_read` but the resulting warning messages are more confusing than helpful. These functions were here for a very long time and soon are going to be removed so let's leave them in peace for a little bit longer. Any bigger modifications, like renaming them, is an unnecessary work.

> Also in the topic branch before everybody starts working on drivers: we need a clear description of _all_ the configuration flags including the ones not touched in this PR, such as `GPIO_INT_DEBOUNCE`

I've clearly stated in the PR and reminded a few times already that the purpose of this PR is to introduce flags used by the Linux DT bindings. Remaining updates to GPIO API are orthogonal to the work done here. This PR alone requires quite a lot of effort to update all the board files, samples and the drivers. Let's focus our energy on one thing and have it done. The GPIO API update took way too much time already. That said, we certainly can start the discussion on how to handle pin debouncing, drive strength configuration in the issue and when the concept is ready update the drivers. But this can and should be done in a separate commit from the one adapting drivers to changes done in this PR and doesn't need to block merging topic-gpio branch into master. I also would welcome having the complete GPIO API updated for 2.1.

> From earlier discussion on future-proofing we should `typedef u32_t gpio_pins_t` and use that for all parameters where the argument encodes a set of pins. The documentation for the type alias should specify that pin `i` is in the set if and only if bit `i` (masked by `1U << i`) is 1. This would allow for a 64-bit pin mask or some more complex encoding if necessary; but more importantly it would make it far more clear to a user how to construct the corresponding argument.

Sounds good, I'll update PR accordingly. 

> I would also like to investigate the opportunities to reduce the number of API functions the driver must implement by delegating more things to the fundamental `set_clear` operator.

If you have an idea how to improve the API that's much welcomed but should be done preferably before we start to update the drivers. I'll wait for the proposal, but if you can't come up with one that is simpler or works faster or generates smaller code size please state so as not to block work on the necessary driver updates.

pabigot commented:
Except for the documented interpretation of the set and clear bits the remaining issues I raised fall under this comment:

> From my perspective they are not blockers for merging this PR, and I'm willing to do the PRs to address them. I just want the thoughts out there so people can think about them.

I remember the slack comment, which I think was an attempt to focus attention on more basic issues, but it's no longer visible in the channel so we can't tell.  The API telecon was later, and appears to be the one on 07-30.  The notes support my recollection that the dedicated set, clear, and toggle functions were expected to be implemented on top of the set_clear basis function.  I'm not asking you to make that change, just to revise the documentation so it's clear that path is allowed.

In any case I definitely do not intend to delay getting this into the topic branch: it's excellent work,  and we need to start working with it.  Please do not wait for anything more from me before pushing an update.

mnkp commented:
> The notes support my recollection that the dedicated set, clear, and toggle functions were expected to be implemented on top of the set_clear basis function. I'm not asking you to make that change, just to revise the documentation so it's clear that path is allowed.

During API meeting we've decided how the public API should look like and mentioned that the driver API should preferably export smaller set of functions to reduce driver code size and ease development. You proposed to have set_clear function as the basis function provided by the driver on top of which remaining public API is build but no decisions were made. The details of the driver API implementation were left open. Understandably, since no one had enough background knowledge to make an informed choice.

Now, looking at the proposed implementation it's clear that the set_clear function is pretty ill suited for the purpose of this API since it's unable to deal efficiently with the differences between logic and physical levels. The conversion is simply too expensive.

pabigot commented:
> Now, looking at the proposed implementation it's clear that the set_clear function is pretty ill suited for the purpose of this API since it's unable to deal efficiently with the differences between logic and physical levels. The conversion is simply too expensive.

Where's the proposed implementation you looked at?  What overhead is "too expensive"?

You may be right, but I don't think simply asserting something makes it true.  If we could get your changes merged so we have something to start from, we could try it and find out.

mnkp commented:
> > the set_clear function is pretty ill suited for the purpose of this API since it's unable to deal efficiently with the differences between logic and physical levels. The conversion is simply too expensive.
> 
> Where's the proposed implementation you looked at? 

The implementation in this PR is the only one that was proposed so far. 

> What overhead is "too expensive"?

The overhead of converting between logical and physical levels. It's there in the sentence you quote.
 
> You may be right, but I don't think simply asserting something makes it true. 

That goes both ways. I've provided an example implementation. It's quite easy to analyse it and think of alternative scenarios. If you believe you can propose some improvements to the current design then by no means please do so. Be specific. But if you keep bringing up some random ideas, like the set_clear function, without ever checking their suitability to solve the problem at hand you are just creating noise. It's quite distracting.

> If we could get your changes merged so we have something to start from, we could try it and find out.

You can test this PR already, there is no need to merge it [Checking out pull requests locally](https://help.github.com/en/articles/checking-out-pull-requests-locally).

mnkp commented on include/drivers/gpio.h:
```diff
@@ -226,13 +557,448 @@ static inline int gpio_pin_write(struct device *port, u32_t pin,
+
+	ret = gpio_port_get_raw(port, value);
+	if (ret == 0) {
+		*value ^= data->invert;
```
@andrewboie could you please double check if accessing driver data directly in the inline function run within the thread context doesn't violate requirements of userspace mode. If that's not OK I'll add an extra function to the driver API to get the value of data.invert, however that would mean longer execution time.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -87,37 +400,42 @@ struct gpio_callback {
+ * Pin with index n, or its value, is identified by bit n (1U << n) in the value
+ * of type gpio_pins_t.
+ */
+typedef u32_t gpio_pins_t;
```
This typedef probably doesn't belong in this file as it is architecture specific. I'm not sure where to put it though.

mnkp commented:
I've pushed a new version with the following changes implemented:
- the old GPIO flags used within DTS and sourced from `include/dt-bindings/gpio/gpio.h` are no longer deprecated. Instead they are deprecated in `include/gpio/gpio.h`. As a result DTS files using old flags will continue to compile and user will see deprecated warnings if the flags are used within the source code.
- introduced `gpio_pins_t` type to be used by the port functions.
- added toggle functionality to `gpio_port_set_clr_bits*` functions.

pabigot committed:
gpio: use named types to distinguish pin sets from pin values

Both pin sets and values encoding pin values are ultimately represented
by 32-bit unsigned integers. Provide typedefs that make the role of a
parameter explicit.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>
Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp commented:
> What's the intended idiom to be used to disable interrupts on a pin that had been configured to enable interrupts, e.g. as a replacement for `gpio_pin_disable_callback()`? Just passing `GPIO_INT_DISABLE` in flags runs afoul of the assert.

That was a bug in the implementation. To disable the interrupt one should pass `GPIO_INT_DISABLE` 
flag to `gpio_pin_interrupt_configure` function, as you have attempted. I've just pushed an update with a fix. Thanks for the report.

mnkp commented:
I've pushed a new version which includes following changes proposed by @pabigot in #18689. 

To keep commit history clean following, relatively minor, updates are incorporated directly into existing commits:
- gpio: document replacement for deprecated callback functions
- gpio: standardize deprecation message style
- drivers/gpio_mcux: implement initial output flags

I cherry picked the following commit:
- gpio: use named types to distinguish pin sets from pin values

mnkp commented:
I've pushed a new version which includes following changes proposed by @pabigot in #18689 incorporated directly into existing commits:
- "gpio: simplify port clear bits"
- "drivers: gpio: fix handler for output parameter"
- removal of toggle functionality from set_clr functions

I have also moved deprecation of gpio_pin_read, gpio_pin_write functions to a separate commit. This makes it easier to deprecate these functions at a later point in time by moving the commit in the commit history, e.g. before merging changes to master.

mnkp committed:
tests: add gpio_api_1pin testcase

gpio_api_1pin testcase verifies following new GPIO API functionality:
- pin active level flags
- input/output configuration flags
- pin drive flags
- gpio_port_*, gpio_pin_* functions
- pin interrupts

The test is using only 1 GPIO pin (defined in DTS as LED0) and relies on
the ability of the driver to configure pin simultanously as in/out.
Drivers that do not allow to configure pins in in/out mode should still
pass the test, however, most of the testcases will be skipped.

The test does not require any modifications to board hardware.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp commented:
I don't like adding new commits to a PR that has been reviewed, however to enforce consistent driver quality I've added "tests: add gpio_api_1pin testcase". It verifies following new GPIO API functionality:
- GPIO pin active level flags
- GPIO input/output configuration flags
- GPIO pin drive flags
- gpio_port_*, gpio_pin_* functions

The test is design to use a single pin, the LED0, which is defined for most of the boards. As such it can be run on the boards as is, without any modifications. There is also test prepared by @pabigot in #18689. This one is using separate pins for input and output. We will need it too.

The gpio_nrfx and gpio_mcux drivers fail the gpio_api_1pin test currently. I pushed the test anyway so it can be reviewed.

pabigot commented:
@mnkp the 1pin test seems fine to me.

I've updated #18689 with a few more changes; the three `FIXUP` patches should be squashed into your commits.   The others can wait, or be cherry-picked, your choice.

There's also a patch that temporarily disables the deprecation warnings, which makes it possible for the  branch to pass sanity check.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -226,13 +988,49 @@ static inline int gpio_pin_write(struct device *port, u32_t pin,
+ * @retval 0 If successful.
+ * @retval -ENOTSUP if any of the configuration options is not supported.
+ * @retval -EINVAL Invalid argument.
+ */
```
Some hardware has limited number of interrupt lines that are shared between different GPIO ports. I will add `-EACCES` "Permission denied" error code to indicate that it's not possible to configure interrupt for a given pin if the hardware resource is already taken.
```
 * @retval -EACCES Interrupt line required to configure pin interrupt is already in use.
```
Another error code that we possibly could use is -EPERM "Operation not permitted".

pabigot replied:
`-EBUSY` "Device or resource busy" would also be appropriate.

mnkp replied:
Yes, that's probably a better choice. I'll use `-EBUSY`.

mnkp commented:
@dbkinder The architectural part of this PR has been discussed and is not expected to change. Could you please review the documentation. Only the two files in the include folder: `include/drivers/gpio.h`, `include/dt-bindings/gpio/gpio.h`.

mnkp commented:
I've pushed new version with following changes:
- Applied "FIXUP: drivers: gpio: fix misuse of u8_t where bool is intended" from #18689.
- Added casting of `port->driver_data` following comment by @pabigot "C++ warns on implicit conversion of void pointers to non-void variables." in #18689.
- Extended gpio_api_1pin testcase to test interrupts.
- Added minor fixes to drivers or gpio_api_1pin testcase to ensure the test is passing on all updated drivers.
- Added `-EBUSY` error code to gpio_pin_interrupt_configure function.

anangl commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -253,6 +255,91 @@ static int gpio_nrfx_read(struct device *port, int access_op,
+		 * ignored as long as the pin is controlled by GPIOTE.
+		 * Attempting to write a pin as a normal GPIO pin will have no
+		 * effect. Controlling the pin via GPIOTE, used to trigger edge
+		 * interrupts, is currently not supported.
```
I think I don't understand the last sentence. What is not supported is configuring the edge triggered interrupt for a pin configured as an output. And that is because for triggering such interrupt we use a GPIOTE channel in Event mode, what effectively configures the pin as an input, overriding the DIR setting in GPIO.

anangl commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -414,7 +506,7 @@ static inline void fire_callbacks(struct device *port, u32_t pins)
 		 * iteration, as some callbacks may get disabled also in any
 		 * of the handlers called here.
```
This comment became outdated.

anangl commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -280,7 +367,7 @@ static int gpio_nrfx_pin_manage_callback(struct device *port,
 
 	for (u8_t curr_pin = from_pin; curr_pin <= to_pin; ++curr_pin) {
-		WRITE_BIT(data->int_en, curr_pin, enable);
+		WRITE_BIT(data->pin_int_en, curr_pin, enable);
```
I have one doubt here. After `int_en` is removed and only `pin_int_en` is used to control the enabling of interrupts, it is possible to enable an interrupt with undefined configuration by calling only `gpio_pin_enable_callback`, and I guess it should not be allowed. Or did I miss something?

anangl replied:
I think we should stay with two variables for now and switch to one when this deprecated function is removed.

mnkp replied:
I agree, moving to one variable was a bit premature. I reverted back to the old design. One difference is that now interrupts are enabled as soon as `gpio_pin_configure` or `gpio_pin_interrupt_configure` function is called - in agreement with the new API design - there is no need to call `gpio_pin_enable_callback` additionally.

anangl commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -321,7 +414,7 @@ static inline u32_t get_level_pins(struct device *port)
 	/* Take into consideration only pins that were configured to
 	 * trigger interrupts and have callback enabled.
```
This comment became outdated.

mnkp commented:
I've pushed a new version:
- Fixes for gpio_nrfx driver as requested by @anangl.
- Removed commits that deprecate legacy functions and flags.
- Added "api: gpio: remove deprecated gpio_port functions" which removes old gpio_port functions which were deprecated in Zephyr release 1.14.
- Some fixes to make Shippable happy. Minor issues, though actual bugs which went unnoticed, since they were masked by multitude of errors reported by deprecated functions and macros.

galak committed:
ci: bump number of minions to see if this builds

Temp patch to see if this at least passes a full shippable build
or not.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
> I've pushed a new version:
> 
> * Fixes for gpio_nrfx driver as requested by @anangl.
> * Removed commits that deprecate legacy functions and flags.
> * Added "api: gpio: remove deprecated gpio_port functions" which removes old gpio_port functions which were deprecated in Zephyr release 1.14.
> * Some fixes to make Shippable happy. Minor issues, though actual bugs which went unnoticed, since they were masked by multitude of errors reported by deprecated functions and macros.

I pushed a temp patch to increase the number of builders for shippable to see if we can get a full shippable run before merging this into the branch.  My suggestion is to let shippable run, we'll remove the commit if all passes and manual merge it into the branch.

galak commented:
> I pushed a temp patch to increase the number of builders for shippable to see if we can get a full shippable run before merging this into the branch. My suggestion is to let shippable run, we'll remove the commit if all passes and manual merge it into the branch.

Looks like to failures:

Building anything for board `nrf52840_pca10090` and `samples/boards/up_squared/gpio_counter/sample.board.up_squared.gpio_counter` on `up_squared`

pabigot commented on include/drivers/gpio.h:
```diff
@@ -193,17 +544,416 @@ static inline int z_impl_gpio_disable_callback(struct device *port,
+static inline int gpio_pin_set(struct device *port, unsigned int pin, int value)
+{
+	struct gpio_driver_data *const data =
+			(struct gpio_driver_data *const)port->driver_data;
```
You're making an immutable pointer to mutable data, which isn't very useful and looks weird because other pointers aren't treated that way.  The fixup I provided made clear that the code should not be changing `invert`:
```
-       struct gpio_driver_data *const data = port->driver_data;
+       const struct gpio_driver_data *data =
+               (const struct gpio_driver_data *)port->driver_data;
```
3 occurrences if you want to change it.

galak commented:
Pulled in https://github.com/zephyrproject-rtos/zephyr/pull/18689/commits/dc2d4c03d1cc6b89d78e6a56b90f5e7b46b4e814 and merged with the NRF commit.
